File: lib/erbx/token.rb
# frozen_string_literal: true

module ERBX
  class Token
    attr_reader :value, :range, :start, :end, :type

    def initialize(value, range, start, end_loc, type)
      @value = value
      @range = range
      @start = start
      @end = end_loc
      @type = type
    end

    def to_hash
      {
        value: value,
        range: range&.to_hash,
        start: start&.to_hash,
        end: self.end&.to_hash,
        type: type
      }
    end

    def to_json(*args)
      to_hash.to_json(*args)
    end

    def inspect
      %(#<ERBX::Token type="#{type}" value="#{value}" range=[#{range.start_position}, #{range.start_position}] start=#{start.line}:#{start.column} end=#{self.end.line}:#{self.end.column}>)
    end
  end
end

File: node/lib/token.ts
import { Range } from './range.js';
import { Location } from './location.js';

export class Token {
  readonly value: string;
  readonly range: Range;
  readonly startLocation: Location;
  readonly endLocation: Location;
  readonly type: string;

  constructor(
    value: string,
    range: Range,
    startLocation: Location,
    endLocation: Location,
    type: string
  ) {
    this.value = value;
    this.range = range;
    this.startLocation = startLocation;
    this.endLocation = endLocation;
    this.type = type;
  }

  toHash(): Record<string, any> {
    return {
      value: this.value,
      range: this.range?.toArray(),
      start_location: this.startLocation?.toHash(),
      end_location: this.endLocation?.toHash(),
      type: this.type,
    };
  }

  toJSON(): Record<string, any> {
    return this.toHash();
  }

  treeInspect(): string {
    return `"${this.value}" (location: ${this.startLocation.treeInspect()}-${this.endLocation.treeInspect()})`;
  }

  inspect(): string {
    return `#<Token type="${this.type}" value=${JSON.stringify(this.value)} range=${this.range.treeInspect()} start=${this.startLocation.treeInspect()} end=${this.endLocation.treeInspect()}>`;
  }

  toString(): string {
    return this.inspect();
  }
}
